### ISSUE TITLE
Memcache: Rails.cache.fetch_multi is slow by an order of magnitude than Rails.cache.read

### ISSUE THREAD

**Issue by sobrinho** [rails/rails#31909]
### Steps to reproduce

Using the bug report template doesn't reproduce the issue for some reason:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "~> 5.1.0"
  gem "dalli"
end

require "rack/test"
require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    cache = ActiveSupport::Cache.lookup_store(:mem_cache_store, ['127.0.0.1:11211'])

    cache.write("key1", true)
    cache.write("key2", true)
    cache.write("key3", true)

    fetch_multi_time = Benchmark.measure do
      10_000.times do
        cache.fetch_multi("key1", "key2", "key3") do |_key|
          raise 'miss'
        end
      end
    end

    read_write_time = Benchmark.measure do
      10_000.times do
        cache.read("key1") or raise 'miss'
        cache.read("key2") or raise 'miss'
        cache.read("key3") or raise 'miss'
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
end

require "minitest/autorun"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    get "/"
    assert last_response.ok?
    raise last_response.body
  end

  private
    def app
      Rails.application
    end
end
```

You have to create a real application, add `dalli` to Gemfile, configure the environment to use mem cache as cache store and create a controller with the same code.

```ruby
class MoviesController < ApplicationController
  def index
    Rails.cache.write("key1", true)
    Rails.cache.write("key2", true)
    Rails.cache.write("key3", true)

    fetch_multi_time = Benchmark.measure do
      1_000.times do
        Rails.cache.fetch_multi("key1", "key2", "key3") do |_key|
          raise 'miss'
        end
      end
    end

    read_write_time = Benchmark.measure do
      1_000.times do
        Rails.cache.read("key1") or raise 'miss'
        Rails.cache.read("key2") or raise 'miss'
        Rails.cache.read("key3") or raise 'miss'
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
end
```

### Expected behavior

I was expecting the `fetch_multi` to perform a lot better than using 3 separated reads.

### Actual behavior

Be aware that using the bug report template does not reproduce the issue (work as expected):

```js
{
  "fetch_multi_time": 2.637000000104308,
  "read_write_time": 3.379054999910295
}
```

But using a fresh application does reproduce (work as not expected):

```javascript
{
  "fetch_multi_time": 0.3341950001195073,
  "read_write_time": 0.01899300003424287
}
```

read + write time is an order of magnitude faster than using fetch multi and this is not supposed to be true.

Note that using dalli directly, the performance is slight better as expected (would be better by having network latency):

```ruby
  def index
    if params[:dalli]
      dalli = Dalli::Client.new('localhost:11211')
      dalli.set("key1", true)
      dalli.set("key2", true)
      dalli.set("key3", true)

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          dalli.get_multi("key1", "key2", "key3")
        end
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          dalli.get("key1")
          dalli.get("key2")
          dalli.get("key3")
        end
      end
    else
      Rails.cache.write("key1", true)
      Rails.cache.write("key2", true)
      Rails.cache.write("key3", true)

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          Rails.cache.fetch_multi("key1", "key2", "key3") do |_key|
            raise 'miss'
          end
        end
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          Rails.cache.read("key1") or raise 'miss'
          Rails.cache.read("key2") or raise 'miss'
          Rails.cache.read("key3") or raise 'miss'
        end
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
```

```js
{
  "fetch_multi_time": 0.23557999986223876,
  "read_write_time": 0.29412800003774464
}
```

So, by that we can assume that the problem is not the Memcached or Dalli driver.

These values are consistent on my machine across several reloads.

We may have some sort of optimization on read calls that aren't present on fetch_multi or something like that, I'm not sure.

### System configuration

**Rails version**: 5.1.6

**Ruby version**: 2.4.2

**Comment by sobrinho** [comment]
One more discovery, dalli store doesn't have the problem, I'm trying to investigate the differences between them.

**Comment by fatkodima** [comment]
Maybe this is due to https://github.com/rails/rails/blob/83c1ed9a1a11196cab66d9c44a56a902ca0710e4/activesupport/lib/active_support/cache/mem_cache_store.rb#L31 used for `#read`?

**Comment by sobrinho** [comment]
Got it.

Repeated calls are faster using read+write because there is an in-memory cache for the read operation but not for the fetch multi operation.

I'm going to try to implement the in-memory cache for this to have the same benefits and open a PR.

**Comment by sobrinho** [comment]
@fatkodima yes, sent my message without reading yours.

I'm going to work on this :)

**Comment by sobrinho** [comment]
Closed in favor of #31911.
